### PR TITLE
fix: #3104 by adding new templates

### LINKED ISSUE #3104: Add FieldErrorTemplate and FieldHelpTemplate
### Prerequisites

- [X] I have read the [documentation](https://react-jsonschema-form.readthedocs.io/)

### What theme are you using?

core

### Is your feature request related to a problem? Please describe.

The `ErrorList` and `Help` components are hidden inside of the `SchemaField` implementation and really need to be exported as new templates so that users (and themes) can customise them.

### Describe the solution you'd like

Add two new templates by refactoring the components out of `SchemaField` in `core`. This will required adding new types for the templates and adding them to the `TemplatesType` in `utils`.

### Describe alternatives you've considered

_No response_

### PR DESCRIPTION
**Pull request by heath-freenome**
### Reasons for making this change

Fixes #3104 as follows:

- Updated the `CHANGELOG.md` appropriately
- Updated the documentation to discuss the new `ArraySchemaField` field and the `FieldErrorTemplate` and `FieldHelpTemplate` templates
- Updated `ArrayField` to get the new `ArraySchemaField` when rendering the `children` of an array item
- Updated `fields/index.ts` to add `ArraySchemaField` as `SchemaField`
- Refactored the `ErrorList` component from inside `SchemaField.tsx` as the new `FieldErrorTemplate`
- Refactored the `Help` component from inside `SchemaField.tsx` as the new `FieldHelpTemplate`
- Updated `SchemaField` to get the `FieldErrorTemplate` and `FieldHelpTemplate` from the `registry`/`uiOptions` using `getTemplate()`
  - Used these templates to render the `errors` and `help` components to the Field
- Updated the tests to verify the new `ArraySchemaField`, `FieldErrorTemplate` and `FieldHelpTemplate` implementations
- Updated the `types.ts` in the `utils` package to add new `FieldErrorProps` and `FieldHelpProps` as well as `FieldErrorTemplate` and `FieldHelpTemplate` to the `TemplatesType`
- Updated the `getTemplates()` tests to add the two new templates
- Added NODE_OPTIONS to the `ci.yml` script to avoid OOM exceptions
- Also, added an additional test for the help/error functionality to all of the theme's Form.test.(j|t)s files

### Checklist

* [x] **I'm updating documentation**
  - [x] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [x] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [x] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Review by jacqueswho — APPROVED** [review]

File: docs/advanced-customization/custom-templates.md [hunk 1]
```diff
@@ -449,6 +451,85 @@ The following props are passed to the `ErrorListTemplate`:
...
+- `schema`: The schema for the field
+- `uiSchema`: The uiSchema for the field
+- `idSchema`: An object containing the id for this field & ids for its properties.
+- `errors`: An array of all errors for this field
+- `errorSchema`: The `ErrorSchema` for this field
+- `registry`: The `Registry` object
+
+## FieldHelpTemplate
+The `FieldErrorTemplate` is the template that renders the all the errors associated a single field.
+If you want to customize how all the errors are rendered you can.
```
Comment (nickgros) [inline docs/advanced-customization/custom-templates.md:502]: Update to refer to FieldHelpTemplate

File: docs/advanced-customization/custom-widgets-fields.md [hunk 2]
```diff
@@ -351,3 +352,32 @@ render((
 If you're curious how this could ever be useful, have a look at the [Kinto formbuilder](https://github.com/Kinto/formbuilder) repository to see how it's used to provide editing capabilities to any form field.
 
 Props passed to a custom SchemaField are the same as [the ones passed to a custom field](#field-props).
+
+### Custom ArraySchemaField
+
+Everything that was mentioned above for a `Custom SchemaField` applies, but this is only used to render the Array item `children` that are then passed to the `ArrayFieldItemTemplate`.
+By default, `ArraySchemaField` assigned to the main `SchemaField` implementation.
```
Comment (nickgros) [inline docs/advanced-customization/custom-widgets-fields.md:359]:
```suggestion
By default, `ArraySchemaField` uses the main `SchemaField` implementation.
```

File: package.json [hunk 3]
```diff
@@ -15,7 +15,7 @@
     "lint": "lerna run lint",
     "cs-check": "lerna run cs-check",
     "cs-format": "lerna run cs-format",
-    "build": "lerna run --stream build",
+    "build": "NODE_OPTIONS=--max-old-space-size=4096 lerna run --concurrency 1 --stream build",
```
Comment (nickgros) [inline package.json:18]: This isn't desirable, it's pretty slow as is. Are you running out of memory locally? If not, these options should only be used in CI.

We could also set the memory higher, linux VMs in GH actions have 7GB RAM and 2 cores.

https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

File: packages/core/test/ArrayField_test.js [hunk 4]
```diff
@@ -2192,7 +2192,7 @@ describe("ArrayField", () => {
         },
       ],
     };
-    it("should pass form context to schema field", () => {
+    it("should pass form context to array schema field only", () => {
```
Comment (nickgros) [inline packages/core/test/ArrayField_test.js:2195]: Shouldn't this be an additional test, rather than changing the existing test?

**Review by nickgros — CHANGES_REQUESTED** [review]

File: packages/core/test/ArrayField_test.js (hunk 4 above)
Comment (heath-freenome) [reply packages/core/test/ArrayField_test.js:2195]: Good point, I'll add that

File: package.json (hunk 3 above)
Comment (heath-freenome) [reply package.json:18]: I'll revert this and try updating the `NODE_OPTIONS` in the `ci.yml`

**Review by nickgros — APPROVED** [review]

**Pull request merged**